---
title: "New mediator genes"
author: "Jinliang Yang"
date: "1-12-2021"
output: NULL
---


## Normalize the path:

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE) 
knitr::opts_knit$set(root.dir=normalizePath('../../')) 
```

#### New results

- Direct SNP: /common/jyanglab/zhikaiyang/projects/mediation/largedata/dsnps_vs_gwas/dsnps_34013rows.csv
- Indirect SNP: /common/jyanglab/zhikaiyang/projects/mediation/largedata/isnps_vs_feature/isnps_47143rows.csv
- Mediators: data/mediators_18812rows_by_cat.csv

### Old results for microbiome data

- cache/microbiome_mediators_34155rows.csv



### Mediators

```{r}
m <- read.csv("data/mediators_18812rows_by_cat.csv")
table(m$method)

### only consider std nitrogen condition
mi <- read.csv("cache/microbiome_mediators_34155rows.csv")

```
- eq: MedFix0.5
- fixed: MedFixBIC

- e2m: p-values for the exposure effect on the individual mediators
- m2y: p-values for the mediator effect on phenotype
- e2m2y: min value among c(1, max(e2m, m2y))

```{r}
# res.eq
m1 <- subset(m, method %in% "res.fixed" & padj <= 0.05)

m2 <- subset(mi, treatment %in% "stdN" & method %in% "res.fixed" & padj <= 0.05)

nrow(m1) # 932
length(unique(m1$id)) # 736

pertrait <- data.frame(table(m1$trait))
mean(pertrait$Freq)


library(plyr)
nt1 <- ddply(m1, .(tissue, category, trait), nrow)
nt1$type <- "Morphology"
nt1[nt1$category == "Metabolite", ]$type <- "Metabolite"
nt1 <- nt1[, -2]

nt2 <- ddply(m2, .(tissue, trait), nrow)
nt2$type <- "Microbiome"

nt <- rbind(nt1, nt2)
write.table(nt, "cache/mediator_summary.csv", sep=",", row.names=FALSE, quote=FALSE)
```

# Number of mediator genes by category and trait

```{r}
library(ggplot2)

fsize=18


med <- read.csv("cache/mediator_summary.csv")
med <- subset(med, type %in% c("Metabolite", "Morphology"))

med $tissue <- factor(med$tissue, levels = c( "GRoot", "GShoot", "L3Base", "L3Tip", "LMAD", "LMAN","Kern"),
                    labels =c( "Germinating Root", "Germinating Shoot", "Leaf 3 base", "Leaf 3 tip", "Adult leaf during the day", "Adult leaf at night","Kernels after pollination"))



pall <- ggplot(med, aes(x=tissue, y=V1, fill=type)) +
    #geom_bar(stat="identity", position=position_dodge()) +
    geom_boxplot(outlier.colour="black", outlier.shape=16, 
             outlier.size=1, notch=FALSE) +
    # scale_fill_manual(values=c("#69b3a2", "#cdc0b0")) +
    # scale_fill_viridis(discrete = TRUE, alpha=0.6) +
    # geom_jitter(color="#cc4141", size=2, alpha=0.9) +
    xlab("") +
    ylab("Avg number of mediator genes") +
    ggtitle("") +
    theme_classic() +
    labs(fill = "") +
    #scale_y_continuous(limits = c(-0.01, 1)) +
    theme(plot.title = element_text(size=20, face = "bold"), 
          axis.text=element_text(size=16, face="bold"),
          strip.text.x = element_text(size = 16, face = "bold"),
          axis.title=element_text(size=fsize, face="bold"),
          axis.text.x = element_text(angle = 15, hjust=0.8),
          legend.position = "none", 
          legend.title = element_text(size=fsize, face="bold"),
          legend.text = element_text(size=fsize))
pall

pdf("graphs/avg_num_mediator_genes_by_type.pdf", width = 10, height= 6)
pall
dev.off()
```







