---
title: "Mediator genes break down by tissue"
author: "Jinliang Yang"
date: "1-21-2021"
output: NULL
---


## Normalize the path:

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE) 
knitr::opts_knit$set(root.dir=normalizePath('../../')) 
```


### Mediators

Note, here, we only focus on m1 (metabolite and morphological traits)
```{r}
library("data.table")
isnp <- fread("/common/jyanglab/zhikaiyang/projects/mediation/largedata/isnps_vs_feature/isnps_47143rows.csv", data.table=FALSE)
# table(isnp$method)

#  res.eq res.fixed
#    22832     24463

### only consider res.fixed
i1 <- subset(isnp, method == "res.fixed")
i1$type <- "Morphology"
idx <- grep("^X\\.", i1$trait)
i1[idx, ]$type <- "Metabolite"

```


# iSNP for FG005

Find the top mediators: "AC148152.3_FG005"


```{r}
ifg5 <- subset(i1, medi %in% "AC148152.3_FG005")

v34 <- fread("cache/gene_v34_conversion.txt", data.table=FALSE)
fg5 <- subset(v34, V1 %in% "AC148152.3_FG005")

# 238143557

chr2 <- subset(ifg5, seq == 2)

chr2$dist <- chr2$Pos - 238143557

out <- subset(chr2, abs(dist) < 100000)
```


```{r}
library(readxl)

bx <- read_excel("data/bx_genes.xlsx")
bx <- as.data.frame(bx)
write.table(bx, "cache/bx_genes.csv", sep=",", row.names = FALSE, quote=FALSE)

bx <- read.csv("data/bx_genes.csv")
bx4 <- merge(v34, bx, by.x="V2", by.y="GeneID")
```


Ask Gen to draw the LD plot


# Find the phenotypic effect

SNP: 238153041

```{r}
library("data.table")
snp <- fread("data/Zm00001d007718.up_down.50K.hmp.txt", data.table=FALSE)

mysnp <- subset(snp, pos ==  238153041)

tsnp <- as.data.frame(t(mysnp))
tsnp$geno <- row.names(tsnp)
tsnp <- tsnp[-(1:11), ]
names(tsnp)[1] <- "S2_238153041"
table(tsnp[,1])
```


```{r}
df <- read.csv("cache/fg005_rna_pheno.csv")
df <- merge(df, tsnp, by.x="genotype", by.y="geno")



```





### Expression


```{r}
library(ggplot2)

fsize=18

# Heatmap 
df1 <- subset(df, trait %in% "EarHeight")
df1 <- subset(df1, S2_238153041 %in% c("AA", "GG"))
pall <- ggplot(df1, aes(x=S2_238153041, y=log(AC148152.3_FG005), fill=S2_238153041)) + 
  geom_violin(trim=FALSE) +
  geom_dotplot(binaxis='y', stackdir='center', dotsize=1)+ 
  stat_compare_means(aes(group = S2_238153041), label = "p.signif") +
    xlab("") +
    ylab("log(TPM)") +
    ggtitle("") +
    theme_classic() +
    labs(fill = "") +
    #scale_fill_manual(values=c("#999999", "#E69F00", "#56B4E9")) +
    theme(plot.title = element_text(size=20, face = "bold"), 
          axis.text=element_text(size=16, face="bold"),
          strip.text.x = element_text(size = 16, face = "bold"),
          axis.title=element_text(size=fsize, face="bold"),
          #axis.text.x = element_text(angle = 15, hjust=0.8),
          #legend.position = c(0.2, 0.8), 
          legend.position = "none",
          legend.title = element_text(size=fsize, face="bold"),
          legend.text = element_text(size=fsize))
pall

pdf("graphs/rna_FG005_AG_alleles.pdf", width = 5, height= 5)
pall
dev.off()
```

```{r}
t.test(log(subset(df1, S2_238153041 %in% "AA")$AC148152.3_FG005), log(subset(df1, S2_238153041 %in% "GG")$AC148152.3_FG005))
```


```{r}
library(ggplot2)
library(agricolae)

fsize=18

# Heatmap 
df1 <- subset(df, trait %in% "EarHeight")
df1 <- subset(df1, S2_238153041 %in% c("AA", "GG"))
df1 = merge(df1, ag_trait[,c(1,39)], by = "genotype")

pall <- ggplot(df1, aes(x=S2_238153041, y=AnthesistoSilkInterval, fill=S2_238153041)) + 
  geom_violin(trim=FALSE) +
  geom_dotplot(binaxis='y', stackdir='center', dotsize=1) +
  stat_compare_means(aes(group = S2_238153041), label = "p.signif") +
    xlab("") +
    ylab("Days") +
    ggtitle("") +
    theme_classic() +
    labs(fill = "") +
    #scale_fill_manual(values=c("#999999", "#E69F00", "#56B4E9")) +
    theme(plot.title = element_text(size=20, face = "bold"), 
          axis.text=element_text(size=16, face="bold"),
          strip.text.x = element_text(size = 16, face = "bold"),
          axis.title=element_text(size=fsize, face="bold"),
          #axis.text.x = element_text(angle = 15, hjust=0.8),
          #legend.position = c(0.2, 0.8), 
          legend.position = "none",
          legend.title = element_text(size=fsize, face="bold"),
          legend.text = element_text(size=fsize))
pall

pdf("graphs/pheno_FG005_AG_alleles.pdf", width = 5, height= 5)
pall
dev.off()


t.test(log(subset(df1, S2_238153041 %in% "AA")$AnthesistoSilkInterval), subset(df1, S2_238153041 %in% "GG")$AnthesistoSilkInterval)

```




```{r}
library(ggplot2)

fsize=18

# Heatmap 
df1 <- subset(df, trait %in% "EarHeight")
df1 <- subset(df1, S2_238153041 %in% c("AA", "GG"))
df1 = merge(df1, ag_trait[,c(1,39)], by = "genotype")

pall <- ggplot(df1, aes(x=log(AC148152.3_FG005), y=AnthesistoSilkInterval, fill=S2_238153041)) + 
  geom_point()+
  geom_smooth(method=lm) +
    xlab("log(TPM)") +
    ylab("Days") +
    ggtitle("") +
    theme_classic() +
    labs(fill = "") +
    #scale_fill_manual(values=c("#999999", "#E69F00", "#56B4E9")) +
    theme(plot.title = element_text(size=20, face = "bold"), 
          axis.text=element_text(size=16, face="bold"),
          strip.text.x = element_text(size = 16, face = "bold"),
          axis.title=element_text(size=fsize, face="bold"),
          #axis.text.x = element_text(angle = 15, hjust=0.8),
          #legend.position = c(0.2, 0.8), 
          legend.position = "none",
          legend.title = element_text(size=fsize, face="bold"),
          legend.text = element_text(size=fsize))
pall

pdf("graphs/pheno_FG005_rna_exp.pdf", width = 5, height= 5)
pall
dev.off()


t.test(log(subset(df1, S2_238153041 %in% "AA")$AnthesistoSilkInterval), log(subset(df1, S2_238153041 %in% "GG")$AnthesistoSilkInterval))

```





