---
title: "Prepare data for Med analysis: learned from Zhikai"
author: "Jinliang Yang"
date: "09-23-2020"
output: NULL
---


## Normalize the path:

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE) 
knitr::opts_knit$set(root.dir=normalizePath('../../')) 
```


### SNP data and PCs

copy from Zhikai's folder!

```{bash}
cp /common/jyanglab/zhikaiyang/projects/mediation-analysis/largedata/geno_282/allchr_bisnp_n282_snpid_maf01_geno2_pruned_NA_0_matrix.txt largedata/geno
cp /common/jyanglab/zhikaiyang/projects/mediation-analysis/largedata/geno_282/allchr_bisnp_n282_snpid_maf01_geno2_pruned.eigenvec largedata
```

###preparing files for population 368
```{bash}
cp /common/jyanglab/zhikaiyang/projects/mediation-analysis/largedata/code2020/traits/tpb_p7msnps_368/All_Merged_1.25M_MAF0.05_AGPV4.eigenvec largedata/All_Merged_1.25M_MAF0.05_AGPV4.eigenvec
cp /common/jyanglab/zhikaiyang/projects/mediation-analysis/largedata/code2020/traits/tpb_p7msnps_368/tpb_p7msnps_368_kern_matrix.txt largedata/geno/tpb_p7msnps_368_kern_matrix.txt
```

###preparing files for population 368
```{r}
library(data.table)

trait=fread("data/blup_368_traits_final_NA.txt", header=T,data.table=F)
geno_order <- fread("largedata/geno_order_368.txt", header=T,data.table=F)
colnames(geno_order)[1] = "genotype"
trait=merge(geno_order, trait, by="genotype")

fwrite(trait[,-2],"data/blup_368_traits_final_NA.txt",sep="\t", na = "NA" ,col.names = T,row.names = F,quote=F)


pc <- fread("largedata/All_Merged_1.25M_MAF0.05_AGPV4.eigenvec", header=F,data.table=F)
pc = merge(geno_order, pc, by.x="genotype", by.y = "V1")
fwrite(pc[,-2],"largedata/All_Merged_1.25M_MAF0.05_AGPV4.eigenvec",sep="\t", na = "NA" ,col.names = F,row.names = F,quote=F)

```

```{bash}
module load plink
cd /common/jyanglab/zhikaiyang/projects/mediation/largedata/geno
##Convert binary file to VCF using plink
plink --bfile hmp321_282_agpv4_maf005_miss03_pruned --recode vcf-iid --out hmp321_282_agpv4_maf005_miss03_pruned

module load perl
perl subs_NA_0.pl
```


#preparation for the new geno data
```{r}

library(data.table)
library(Ropt)

trait=fread("trait_GDDDaystoSilk.txt", header=T,data.table=F)
#pheno <- fread("/common/jyanglab/zhikaiyang/projects/mediation-analysis/largedata/geno_282/phenotype_282_GEMMA_sure.txt")
#trait$GDDDaystoSilk <- as.vector(unlist(pheno[,12]))
d=fread("hmp321_282_agpv4_maf005_miss03_pruned.vcf1", header=F,data.table=F)
d1=t(d)
mk=colnames(d1)=d1[1,]
colnames(d1)[1]="Genotype"
d1=d1[-1,]
d2=as.data.frame(d1,stringsAsFactors =F)
d3=as.data.frame(d1)
id = which(d3$Genotype == "W64a")
d3$Genotype = as.character(d3$Genotype)
d3$Genotype[id] = "W64A"
d4=merge(trait, d3, by="Genotype")
out="gddtosilk_p7msnps_matrix_new.txt"
fwrite(d4,"gddtosilk_p7msnps_matrix_new.txt",sep="\t",col.names = T,row.names = F,quote=F)



pc <- fread("largedata/geno/hmp321_282_agpv4_maf005_miss03_pruned.eigenvec", data.table = FALSE)
id = which(pc$V1 == "W64a")
pc$V1[id] = "W64A"
pc$V2[id] = "W64A"
fwrite(pc,"largedata/geno/hmp321_282_agpv4_maf005_miss03_pruned.eigenvec",sep="\t",col.names = F, row.names = F, quote=F)
```

-----
ignore the below code!

```{r}
p <- fread("/common/jyanglab/zhikaiyang/projects/mediation-analysis/largedata/geno_282/allchr_bisnp_n282_snpid_maf01_geno2_pruned_NA_0_matrix.txt", data.table=FALSE)
dim(p) #274 696549
p <- p[!is.na(p$GDDDaystoSilk),] 


pc <- fread("/common/jyanglab/zhikaiyang/projects/mediation-analysis/largedata/geno_282/allchr_bisnp_n282_snpid_maf01_geno2_pruned.eigenvec", data.table = FALSE)
dim(pc) #277

X2 <- merge(pc[, 2:4], X[, 1:3],  by.x="V2", by.y="HMP32Name")
df <- merge(X2, p[, 1:3], by.x="V2", by.y="Genotype")
dim(df)#250
df <- df[order(df$V2),]
gid <- df$V2


p <- p[order(p$Genotype),]            
y <- p[p$Genotype %in% gid, 1:2]
sum(y$Genotype != gid) # must be 0
y <- as.matrix(y[,-1])

row.names(y) <- gid

### Z
Z <- as.matrix(p[p$Genotype %in% gid, -1:-2])
row.names(Z) <- gid

### X
X <- X[order(X$HMP32Name),]
X <- X[X$HMP32Name %in% gid, ]
sum(X$HMP32Name != gid) # must be 0
X <- as.matrix(X[, -1])
row.names(X) <- gid

### X0
pc <- pc[order(pc$V2), ]
pc <- pc[pc$V2 %in% gid, -1]
sum(pc$V2 != gid) # must be 0
X0 <- pc[, 2:4]
row.names(X0) <- gid

###remove genes with expression correlation with phenotype lower than 0.05
cors<-apply(X, 2, function(X) cor(X, y))
X  <- X[, -(which((cors<=0.05) & (cors>=-0.05) ))] #255 12570

library(dplyr)
X <- as.data.frame(X)
X <- mutate_all(X, scale)
row.names(X) <- gid
X <- as.matrix(X)

X0 <- as.matrix(X0)



fwrite(as.list(colnames(X)), "largedata/geno/colnamesX_gs.csv", sep=",", row.names=FALSE, quote=FALSE)
fwrite(as.list(colnames(Z)), "largedata/geno/colnamesZ_gs.csv", sep=",", row.names=FALSE, quote=FALSE)


```


```{bash}
########### New Genotype for simulation study #############

plink -bfile hmp321_282_agpv4_maf005_miss03_pruned --indep-pairwise 100 kb 10 0.005 --out hmp321_282_agpv4_maf005_miss03_pruned_further
#769690 to 
plink -bfile hmp321_282_agpv4_maf005_miss03_pruned --extract hmp321_282_agpv4_maf005_miss03_pruned_further.prune.in --make-bed --out hmp321_282_agpv4_maf005_miss03_pruned_further

plink --bfile hmp321_282_agpv4_maf005_miss03_pruned_further --recode vcf-iid --out hmp321_282_agpv4_maf005_miss03_pruned_further

module load perl
perl subs_NA_0.pl

```




#preparation for the new geno data
```{r}

library(data.table)
library(Ropt)

trait=fread("trait_GDDDaystoSilk.txt", header=T,data.table=F)
#pheno <- fread("/common/jyanglab/zhikaiyang/projects/mediation-analysis/largedata/geno_282/phenotype_282_GEMMA_sure.txt")
#trait$GDDDaystoSilk <- as.vector(unlist(pheno[,12]))
d=fread("hmp321_282_agpv4_maf005_miss03_pruned_further.vcf1", header=F,data.table=F)
d1=t(d)
mk=colnames(d1)=d1[1,]
colnames(d1)[1]="Genotype"
d1=d1[-1,]
d2=as.data.frame(d1,stringsAsFactors =F)
d3=as.data.frame(d1)
id = which(d3$Genotype == "W64a")
d3$Genotype = as.character(d3$Genotype)
d3$Genotype[id] = "W64A"
d4=merge(trait, d3, by="Genotype")
fwrite(d4,"gddtosilk_22ksnps_matrix_new.txt",sep="\t",col.names = T,row.names = F,quote=F)

```


