---
title: "Master Med genes: tissue specificity"
author: "Jinliang Yang"
date: "09-18-2020"
output: NULL
---


## Normalize the path:

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE) 
knitr::opts_knit$set(root.dir=normalizePath('../../')) 
```

### read in data

The 16 candidate genes

```{r}
mg <- read.csv("data/med_master_table_fixed.csv")
### flowering time and Female Inflorescence
f <- subset(mg, trait %in% c("asi", "daystosilk", "daystotassel", "tpb", "tsl"))
f1 <- subset(f, tissue %in% "ln268")
```

The master FPM file

```{r}
# /common/jyanglab/shared/dbcenter/Kremling_Nature3RNASeq282_March2018/Expression_matrix/df_STAR_HTSeq_counts_B73_match_based_on_genet_dist_DESeq2_normed_fpm_rounded_origNames_and_Altnames.txt
library("data.table")
rna <- fread("/common/jyanglab/shared/dbcenter/Kremling_Nature3RNASeq282_March2018/Expression_matrix/df_STAR_HTSeq_counts_B73_match_based_on_genet_dist_DESeq2_normed_fpm_rounded_origNames_and_Altnames.txt", header=TRUE, data.table=FALSE)
# 1960 37136

idx <- which(names(rna) %in% f1$gene3)
myr <- rna[, c(1:9, idx)]

write.table(myr, "cache/goi.csv", sep=",", row.names = FALSE, quote=FALSE)
```

------------------

Back to the local computer!


```{r}
library("data.table")

goi <- fread("cache/goi.csv", header=TRUE, data.table=FALSE)
table(goi$TissueWODate)
```

Compute average expression level by tissue
```{r}
library(tidyr)
library(plyr)

sub <- goi[, c(3, 10:25)]
df <- gather(sub, key="gene", value="fpm", 2:ncol(sub))
dim(subset(df, TissueWODate %in% "LMAD" & gene %in% "AC192244.3_FG007"))
# 210   3

# average gene expression across 200 lines by tissue type
avg <- ddply(df, .(gene , TissueWODate), summarise,
                  avg = mean(fpm, na.rm=TRUE))

```


# Plot the results

```{r}
library(ggplot2)

avg$TissueWODate <- factor(avg$TissueWODate, levels = c( "GRoot", "GShoot", "L3Base", "L3Mid", "L3Tip", "LMAD", "LMAN", "Kern"),
                    labels =c( "Germinating Root", "Germinating Shoot", "Leaf 3 base","Leaf 3 Middle", "Leaf 3 tip", "Adult leaf during the day", "Adult leaf at night","Kernels after pollination"))

## gene list from previous list
glist <- out[order(out$pval),]
glist$id <- gsub("\\..", "", glist$id1)

# df2 <- subset(avg, gene %in% c("GRMZM2G140559", "GRMZM2G054225", "RMZM2G037048", "GRMZM2G020148", "GRMZM2G000231"))

df2 <- subset(avg, gene %in% glist$id[1:6])
p <- ggplot(df2, aes(x=TissueWODate, y=log10(avg), group=gene, color=gene)) +
    geom_line() + 
    geom_point() +
    xlab("") +
    ylab("log10(FPM)") +
    ggtitle("") +
    theme_classic() +
    labs(color = "") +
    #scale_y_continuous(limits = c(-0.01, 1)) +
    #geom_vline(xintercept=0, linetype="dashed", color = "red") +
    theme(plot.title = element_text(size=20, face = "bold"), 
          axis.text=element_text(size=16, face="bold"),
          strip.text.x = element_text(size = 16, face = "bold"),
          axis.title=element_text(size=18, face="bold"),
          #axis.text.x = element_text(angle = 20, hjust=0.6),
          legend.position = c(0.8,0.9), 
          axis.text.x = element_text(angle = 20, hjust=0.6),
          legend.title = element_text(size=18, face="bold"),
          legend.text = element_text(size=18))
p


pdf("graphs/fpm_by_tissue.pdf", width = 8, height= 6)
p
dev.off()
```


#Metabolite: The master FPM file

```{r}
# /common/jyanglab/shared/dbcenter/Kremling_Nature3RNASeq282_March2018/Expression_matrix/df_STAR_HTSeq_counts_B73_match_based_on_genet_dist_DESeq2_normed_fpm_rounded_origNames_and_Altnames.txt
library("data.table")
rna <- fread("/common/jyanglab/shared/dbcenter/Kremling_Nature3RNASeq282_March2018/Expression_matrix/df_STAR_HTSeq_counts_B73_match_based_on_genet_dist_DESeq2_normed_fpm_rounded_origNames_and_Altnames.txt", header=TRUE, data.table=FALSE)
# 1960 37136

idx <- which(names(rna) %in% t5_m_gshoot[(t5_m_gshoot$Freq >= 1),]$Var1)
myr <- rna[, c(1:9, idx)]

write.table(myr, "cache/goi_m.csv", sep=",", row.names = FALSE, quote=FALSE)
```

------------------

#Metabolite: Back to the local computer!


```{r}
library("data.table")

goi_m <- fread("cache/goi_m.csv", header=TRUE, data.table=FALSE)
table(goi_m$TissueWODate)
```
#Metabolite: Compute average expression level by tissue
```{r}
library(tidyr)
library(plyr)

sub <- goi_m[, c(3, 10:146)]
df <- gather(sub, key="gene", value="fpm", 2:ncol(sub))
dim(subset(df, TissueWODate %in% "LMAD" & gene %in% "AC192244.3_FG007"))
# 210   3

# average gene expression across 200 lines by tissue type
avg <- ddply(df, .(gene , TissueWODate), summarise,
                  avg = mean(fpm, na.rm=TRUE))

```


##Metabolite:  Plot the results

```{r}
library(ggplot2)

avg$TissueWODate <- factor(avg$TissueWODate, levels = c( "GRoot", "GShoot", "L3Base", "L3Mid", "L3Tip", "LMAD", "LMAN", "Kern"),
                    labels =c( "Germinating Root", "Germinating Shoot", "Leaf 3 base","Leaf 3 Middle", "Leaf 3 tip", "Adult leaf during the day", "Adult leaf at night","Kernels after pollination"))

## gene list from previous list

df2 <- subset(avg, gene %in% mt_gshoot$id[1:6])
p <- ggplot(df2, aes(x=TissueWODate, y=log10(avg), group=gene, color=gene)) +
    geom_line() + 
    geom_point() +
    xlab("") +
    ylab("log10(FPM)") +
    ggtitle("") +
    theme_classic() +
    labs(color = "") +
    #scale_y_continuous(limits = c(-0.01, 1)) +
    #geom_vline(xintercept=0, linetype="dashed", color = "red") +
    theme(plot.title = element_text(size=20, face = "bold"), 
          axis.text=element_text(size=16, face="bold"),
          strip.text.x = element_text(size = 16, face = "bold"),
          axis.title=element_text(size=18, face="bold"),
          #axis.text.x = element_text(angle = 20, hjust=0.6),
          legend.position = "right", 
          axis.text.x = element_text(angle = 90, hjust=0.6),
          legend.title = element_text(size=18, face="bold"),
          legend.text = element_text(size=18))
p


pdf("graphs/fpm_by_tissue_metabolites.pdf", width = 8, height= 6)
p
dev.off()
```

