---
title: "Indirect SNPs"
author: "Jinliang Yang"
date: "09-20-2020"
output: NULL
---


## Normalize the path:

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE) 
knitr::opts_knit$set(root.dir=normalizePath('../../')) 
```

### Summarize indirect SNPs

From Zhikai:
"/common/jyanglab/zhikaiyang/projects/mediation-analysis/largedata/code2020/traits"
they are in each trait folder
files names ending by indirect_snps.csv
or dsnps.csv



```{r}
# /common/jyanglab/shared/dbcenter/Kremling_Nature3RNASeq282_March2018/Expression_matrix/df_STAR_HTSeq_counts_B73_match_based_on_genet_dist_DESeq2_normed_fpm_rounded_origNames_and_Altnames.txt
library("data.table")

files <- list.files(path="/common/jyanglab/zhikaiyang/projects/mediation-analysis/largedata/code2020/traits",
                   pattern="indirect_snps",recursive = TRUE, full.names=T)

p7m <- files[grep("p7msnps", files)]


out <- data.frame()
for(i in 1:length(p7m)){
  f <- fread(p7m[i], data.table=FALSE, header=TRUE)
  f$file <- p7m[i]
  out <- rbind(out, f)
}

out$file <- gsub(".*\\/", "", out$file)
out$trait <- gsub("_p7msnps_.*", "", out$file)
out$method <- gsub("_.*", "", out$trait)
out$trait <- gsub(".*_", "", out$trait)

out$tissue <- gsub(".*_p7msnps_", "", out$file)
out$tissue <- gsub("_.*", "", out$tissue)

table(out$tissue)
fwrite(out, "cache/indirect_SNPs_p7msnps.csv", sep=",", row.names=FALSE, quote=FALSE)
```

------------------

Back to the local computer!


```{r}
library("data.table")

id <- fread("cache/indirect_SNPs_p7msnps.csv", header=TRUE, data.table=FALSE)

# Unique mediators
length(unique(id$medi)) #447

tb <- data.frame(table(id$snps_for_medi))
tb <- tb[order(tb$Freq, decreasing = T),]
sub <- subset(tb, Freq > 1)
dim(sub)
```


### Quick Manhattan plot

- Zhi

```{r}
library("huskeR")


tb$Var1 <- gsub("-", "_", tb$Var1)
tb$chr <- as.numeric(as.character(gsub("_.*", "", tb$Var1)))
tb$pos <- as.numeric(as.character(gsub(".*_", "", tb$Var1)))

png("graphs/indirect_freq_plot.png", width=11, height=4.25, units="in", res=300)
plot_mht(inputdf=tb, col2plot="Freq", jmph=FALSE, cl_file="data/chr_length_B73v4.csv", pl=TRUE,
         ylab="Indirect SNPs Freq",  cex = 0.6)
dev.off()
```


### Check the most freq indirect SNPs

```{r}
tb <- data.frame(table(id$snps_for_medi))
tb <- tb[order(tb$Freq, decreasing = T),]


med_id3 <- subset(id, snps_for_medi %in% tb$Var1[3])

med_id1 <- subset(id, snps_for_medi %in% tb$Var1[1])
med_id2 <- subset(id, snps_for_medi %in% tb$Var1[2])

med_id4 <- subset(id, snps_for_medi %in% tb$Var1[4])


med_id5 <- subset(id, snps_for_medi %in% tb$Var1[5])
med_id6 <- subset(id, snps_for_medi %in% tb$Var1[6])

med_id7 <- subset(id, snps_for_medi %in% tb$Var1[7])

med_id7 <- subset(id, snps_for_medi %in% tb$Var1[7])

med_top10 <- subset(id, snps_for_medi %in% tb$Var1[1:10])


```


```{r}

mg <- read.csv("data/med_master_table_fixed.csv")

mgt <- data.frame(table(mg$gene3))
mgt <- mgt[order(mgt$Freq, decreasing = T),]

hlgene <- subset(mg, gene3 %in% mgt$Var1[1])

### from the top10 indirect SNPs, which ones contain the FG008
fg008 <- subset(med_top10, medi %in% hlgene$gene3)
### deduplicated
fg008 <- fg008[!duplicated(fg008$snps_for_medi),]

fg008$chr <- as.numeric(as.character(gsub("-.*", "", fg008$snps_for_medi)))
fg008$pos <- as.numeric(as.character(gsub(".*-", "", fg008$snps_for_medi)))

cl <- read.csv("data/chr_length_B73v4.csv")
fg008 <- newpos(d=fg008, GAP=5e+06, cl)

names(hlgene)[4] <- "pos"
hlgene <- newpos(d=hlgene, GAP=5e+06, cl)

fg008$end <- 550111694
names(fg008)[9] <- "start"

tb$Var1 <- gsub("-", "_", tb$Var1)
tb$chr <- as.numeric(as.character(gsub("_.*", "", tb$Var1)))
tb$pos <- as.numeric(as.character(gsub(".*_", "", tb$Var1)))



out = data.frame(start=0, end=120616133)
plot_mht_ellipse(inputdf=tb, edf=fg008, col2plot="Freq", jmph=FALSE, cl_file="data/chr_length_B73v4.csv", pl=TRUE,
         ylab="Indirect SNPs Freq",  cex = 0.6)



```








###   Metabolites   ###

```{r}
library("data.table")

files <- list.files(path="/common/jyanglab/zhikaiyang/projects/mediation/largedata/med_output",
                   pattern="indirect_X\\.",recursive = TRUE, full.names=T)

out <- data.frame()
for(i in 1:length(files)){
  f <- fread(files[i], data.table=FALSE, header=TRUE)
  f$file <- files[i]
  out <- rbind(out, f)
}


out$file <- gsub(".*\\/", "", out$file)
out$trait <- gsub(".*X\\.[0-9]+\\.\\.", "", out$file)
out$trait <- gsub("\\.\\.[0-9]+\\.[0-9]+_ts.*", "", out$trait)

out$method <- gsub("_.*", "", out$file)

out$tissue <- gsub(".*_ts_", "", out$file)
out$tissue <- gsub("\\.csv", "", out$tissue)

table(out$tissue)
fwrite(out, "cache/indirect_SNPs_m.csv", sep=",", row.names=FALSE, quote=FALSE)
```


Back to the local computer!


```{r}
library("data.table")

id <- fread("cache/indirect_SNPs_m.csv", header=TRUE, data.table=FALSE)
id <- subset(id, method == "res.fixed")
id$trait <- gsub("\\.\\.s100.*", "", id$trait)
# Unique mediators
length(unique(id$medi)) #892

tb <- data.frame(table(id$snps_for_medi))
tb <- tb[order(tb$Freq, decreasing = T),]
sub <- subset(tb, Freq > 1)
dim(sub)

```


### Quick Manhattan plot

- Zhi

```{r}
library("huskeR")


tb$Var1 <- gsub("-", "_", tb$Var1)
tb$chr <- as.numeric(as.character(gsub("_.*", "", tb$Var1)))
tb$pos <- as.numeric(as.character(gsub(".*_", "", tb$Var1)))

png("graphs/indirect_freq_plot_m.png", width=11, height=4.25, units="in", res=300)
plot_mht(inputdf=tb, col2plot="Freq", jmph=FALSE, cl_file="data/chr_length_B73v4.csv", pl=TRUE,
         ylab="Indirect SNPs Freq",  cex = 0.6)
dev.off()
```


### Check the most freq indirect SNPs

```{r}
tb <- data.frame(table(id$snps_for_medi))
tb <- tb[order(tb$Freq, decreasing = T),]


med_id3 <- subset(id, snps_for_medi %in% tb$Var1[3])

med_id1 <- subset(id, snps_for_medi %in% tb$Var1[1])
med_id2 <- subset(id, snps_for_medi %in% tb$Var1[2])

med_id4 <- subset(id, snps_for_medi %in% tb$Var1[4])


med_id5 <- subset(id, snps_for_medi %in% tb$Var1[5])
med_id6 <- subset(id, snps_for_medi %in% tb$Var1[6])

med_id7 <- subset(id, snps_for_medi %in% tb$Var1[7])

med_id7 <- subset(id, snps_for_medi %in% tb$Var1[7])

med_top10 <- subset(id, snps_for_medi %in% tb$Var1[1:10])

tb_top10 <- data.frame(table(med_top10$medi))
tb_top10 <- tb_top10[order(tb_top10$Freq, decreasing = T),]


bx13 <- subset(id, medi == "AC148152.3_FG005")
```



###   Microbiome   ###

```{r}
# /common/jyanglab/shared/dbcenter/Kremling_Nature3RNASeq282_March2018/Expression_matrix/df_STAR_HTSeq_counts_B73_match_based_on_genet_dist_DESeq2_normed_fpm_rounded_origNames_and_Altnames.txt
library("data.table")

files <- list.files(path="/common/jyanglab/zhikaiyang/projects/mediation/largedata/med_output", pattern="fixed_indirect_T[0-9]",recursive = TRUE, full.names=T)

#p7m <- files[grep("p7msnps", files)]
p7m <- files

out <- data.frame()
for(i in 1:length(p7m)){
  f <- fread(p7m[i], data.table=FALSE, header=TRUE)
  f$file <- p7m[i]
  out <- rbind(out, f)
}

out$file <- gsub(".*\\/", "", out$file)

out$trait  <- gsub(".*indirect_|_ts.*", "", out$file)
out$tissue  <- gsub(".*_ts_|.csv", "", out$file)
out$method <- gsub("res\\.|_.*", "", out$file)


table(out$tissue)
fwrite(out, "largedata/indirect_SNPs_p7msnps_microbiome_fixed.csv", sep=",", row.names=FALSE, quote=FALSE)
```


Back to the local computer!


```{r}
library("data.table")

id <- fread("largedata/indirect_SNPs_p7msnps_microbiome_fixed.csv", header=TRUE, data.table=FALSE)
id <- subset(id, method == "fixed")
# Unique mediators
length(unique(id$medi)) #892

tb <- data.frame(table(id$snps_for_medi))
tb <- tb[order(tb$Freq, decreasing = T),]
sub <- subset(tb, Freq > 1)
dim(sub)

```


### Quick Manhattan plot

- Zhi

```{r}
library("huskeR")


tb$Var1 <- gsub("-", "_", tb$Var1)
tb$chr <- as.numeric(as.character(gsub("_.*", "", tb$Var1)))
tb$pos <- as.numeric(as.character(gsub(".*_", "", tb$Var1)))

png("graphs/indirect_freq_plot_microbiome.png", width=11, height=4.25, units="in", res=300)
plot_mht(inputdf=tb, col2plot="Freq", jmph=FALSE, cl_file="data/chr_length_B73v4.csv", pl=TRUE,
         ylab="Indirect SNPs Freq",  cex = 0.6)
dev.off()
```


### Check the most freq indirect SNPs

```{r}
tb <- data.frame(table(id$snps_for_medi))
tb <- tb[order(tb$Freq, decreasing = T),]

hist(tb$Freq, main = "Distribution of Deteced Freq of Indirect SNPs", xlab = "Deteced Freq of Indirect SNPs" )

med_id3 <- subset(id, snps_for_medi %in% tb$Var1[3])

med_id1 <- subset(id, snps_for_medi %in% tb$Var1[1])
med_id2 <- subset(id, snps_for_medi %in% tb$Var1[2])

med_id4 <- subset(id, snps_for_medi %in% tb$Var1[4])


med_id5 <- subset(id, snps_for_medi %in% tb$Var1[5])
med_id6 <- subset(id, snps_for_medi %in% tb$Var1[6])

med_id7 <- subset(id, snps_for_medi %in% tb$Var1[7])

med_id7 <- subset(id, snps_for_medi %in% tb$Var1[7])

med_top10 <- subset(id, snps_for_medi %in% tb$Var1[1:10])

tb_top10 <- data.frame(table(med_top10$medi))
tb_top10 <- tb_top10[order(tb_top10$Freq, decreasing = T),]


bx13 <- subset(id, medi == "AC148152.3_FG005")
bx13 = subset(mgmi, id == "AC148152.3_FG005")
bx13t = data.frame(table(bx13$trait))
```

