---
title: "Master Med genes: top mediator genes"
author: "Jinliang Yang"
date: "09-22-2020"
output: NULL
---


## Normalize the path:

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE) 
knitr::opts_knit$set(root.dir=normalizePath('../../')) 
```

### read in data

select the top mediators!

```{r}
mg <- read.csv("data/med_master_table_fixed.csv")

## freq of the mediators
mgt <- data.frame(table(mg$gene3))
mgt <- mgt[order(mgt$Freq, decreasing = T),]
```

The master FPM file

```{r}
# /common/jyanglab/shared/dbcenter/Kremling_Nature3RNASeq282_March2018/Expression_matrix/df_STAR_HTSeq_counts_B73_match_based_on_genet_dist_DESeq2_normed_fpm_rounded_origNames_and_Altnames.txt
library("data.table")
rna <- fread("/common/jyanglab/shared/dbcenter/Kremling_Nature3RNASeq282_March2018/Expression_matrix/df_STAR_HTSeq_counts_B73_match_based_on_genet_dist_DESeq2_normed_fpm_rounded_origNames_and_Altnames.txt", header=TRUE, data.table=FALSE)
# 1960 37136

### select the top 4 with Freq >2
idx <- which(names(rna) %in% mgt$Var1[1:4])
myr <- rna[, c(1:9, idx)]

write.table(myr, "cache/top4_mediators.csv", sep=",", row.names = FALSE, quote=FALSE)
```

------------------

Back to the local computer!


```{r}
library("data.table")

top4 <- fread("cache/top4_mediators.csv", header=TRUE, data.table=FALSE)
table(top4$TissueWODate)
```

Compute average expression level by tissue
```{r}
library(tidyr)
library(plyr)

sub <- top4[, c(3, 10:13)]
df <- gather(sub, key="gene", value="fpm", 2:ncol(sub))
dim(subset(df, TissueWODate %in% "LMAD" & gene %in% "AC148152.3_FG008"))
# 210   3

# average gene expression across 200 lines by tissue type
avg <- ddply(df, .(gene , TissueWODate), summarise,
                  avg = mean(fpm, na.rm=TRUE))

```


# Plot the results

```{r}
library(ggplot2)

avg$TissueWODate <- factor(avg$TissueWODate, levels = c( "GRoot", "GShoot", "L3Base", "L3Mid", "L3Tip", "LMAD", "LMAN", "Kern"),
                    labels =c( "Germinating Root", "Germinating Shoot", "Leaf 3 base","Leaf 3 Middle", "Leaf 3 tip", "Adult leaf during the day", "Adult leaf at night","Kernels after pollination"))

# df2 <- subset(avg, gene %in% c("GRMZM2G140559", "GRMZM2G054225", "RMZM2G037048", "GRMZM2G020148", "GRMZM2G000231"))


p <- ggplot(avg, aes(x=TissueWODate, y=log10(avg), group=gene, color=gene)) +
    geom_line() + 
    geom_point() +
    xlab("") +
    ylab("log10(FPM)") +
    ggtitle("") +
    theme_classic() +
    labs(color = "") +
    scale_y_continuous(limits = c(-1, 3)) +
    #geom_vline(xintercept=0, linetype="dashed", color = "red") +
    theme(plot.title = element_text(size=20, face = "bold"), 
          axis.text=element_text(size=16, face="bold"),
          strip.text.x = element_text(size = 16, face = "bold"),
          axis.title=element_text(size=18, face="bold"),
          #axis.text.x = element_text(angle = 20, hjust=0.6),
          legend.position = c(0.8, 0.95), 
          axis.text.x = element_text(angle = 20, hjust=0.6),
          legend.title = element_text(size=18, face="bold"),
          legend.text = element_text(size=18))
p


pdf("graphs/top4_fpm_by_tissue.pdf", width = 8, height= 6)
p
dev.off()
```

# gene expression and functional study

```{r}
mt <- read.csv("data/function/metabolite_rep1.csv")

### merge tissue by tissue
tem1 <- data.frame(table(top4$TissueWODate))

ts <- merge(top4, mt, by.x="RawPhenotypeNames", by.y="Genotype")
tem2 <- data.frame(table(ts2$TissueWODate))


cortest <- function(ts, mode=1){
  ## do tissue by tissue
  t <- data.frame(table(ts$TissueWODate))
  out <- data.frame()
  for(i in 1:nrow(t)){
    sub <- subset(ts, TissueWODate %in% t$Var1[i] )
    for(j in 14:ncol(ts)){
      if(mode == 1){
        # exp vs. metabolite
        pval <- cor.test(sub[, 11], sub[,j])$p.value
      }else if(mode ==2){
        # log(exp) vs. metabolite
        pval <- cor.test(log(sub[, 11]), sub[,j])$p.value
      }else if(mode ==3){
        # exp vs. log(metabolite)
        pval <- cor.test(sub[, 11], log(sub[,j]))$p.value
      }else if(mode == 4){
        # log(exp) vs. log(metabolite)
        pval <- cor.test(log(sub[, 11]), log(sub[,j]))$p.value
      }
      
      tem <- data.frame(gene=names(sub)[11], metabolite=names(sub)[j], pval=pval, tissue = t$Var1[i])
      out <- rbind(out, tem)
    }
  }
  out$qval <- p.adjust(out$pval, method="fdr")
  out <- out[order(out$qval, decreasing = F),]
  message(sprintf("###>>> [ %s ] metabolites < 0.05", nrow(subset(out, qval < 0.05))))
  return(out)
}

# exp vs. metabolite
outdf1 <- cortest(ts, mode=1)
###>>> [ 1 ] metabolites < 0.05

# log(exp) vs. metabolite
outdf2 <- cortest(ts, mode=2)
###>>> [ 8 ] metabolites < 0.05

# exp vs. log(metabolite)
outdf3 <- cortest(ts, mode=3)
###>>> [ 0 ] metabolites < 0.05

# log(exp) vs. log(metabolite)
outdf4 <- cortest(ts, mode=4)
###>>> [ 6 ] metabolites < 0.05



```


# Plot relationship with metabolite data

```{r}
library(ggplot2)

sub <- subset(ts, TissueWODate %in% "L3Tip" )
p2 <- ggplot(sub, aes(x=log(AC148152.3_FG008), y=X.5960..aspartic.acid.2..13.207.)) +
    geom_point() +
    geom_smooth(method=lm) +
    xlab("log(AC148152.3_FG008)") +
    ylab("aspartic acid") +
    ggtitle("") +
    theme_classic() +
    #labs(color = "") +
    #scale_y_continuous(limits = c(-0.01, 1)) +
    #geom_vline(xintercept=0, linetype="dashed", color = "red") +
    theme(plot.title = element_text(size=20, face = "bold"), 
          axis.text=element_text(size=16, face="bold"),
          strip.text.x = element_text(size = 16, face = "bold"),
          axis.title=element_text(size=18, face="bold"),
          #axis.text.x = element_text(angle = 20, hjust=0.6),
          legend.position = "none", 
          #axis.text.x = element_text(angle = 20, hjust=0.6),
          legend.title = element_text(size=18, face="bold"),
          legend.text = element_text(size=18))
p2


pdf("graphs/regression_aspacid_FPM.pdf", width = 8, height= 6)
p2
dev.off()
```


```{r}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("snpStats")
devtools::install_github("yangjl/LDheatmap")
```

