---
title: "Collect mediation analysis results"
author: "Zhikai Yang & Jinliang Yang"
date: "10-20-2020"
output: NULL
---


## Normalize the path:

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE) 
knitr::opts_knit$set(root.dir=normalizePath('../../')) 
```

### Results generated by Zhikai

/common/jyanglab/zhikaiyang/projects/mediation/largedata/med_output

fixed_dsnps : direct snps
fixed_indirect : indirect snps
fixed_mediator_trait : mediator for agronomic traits
fixed_mediator_X...: mediator for metabolic traits
fixe.pcut_sum : pvm


### Mediators

```{r}

f <- list.files(path="/common/jyanglab/zhikaiyang/projects/mediation/largedata/med_output/", pattern="mediator_", full.names = T)

out <- data.frame()
for(i in 1:length(f)){
  tem <- read.csv(f[i], header=TRUE)
  tem$file <- f[i]
  out <- rbind(out, tem)
}

out$file  <- gsub(".*\\/", "", out$file)
# method
out$method <- gsub("_.*", "", out$file)
# tissue type
out$tissue <- gsub(".*_ts_|.csv", "", out$file)
table(out$tissue)
# trait
out$trait <- gsub(".*mediator_|_ts.*", "", out$file)
table(out$trait)

write.table(out, "cache/mediators_23296rows.csv", sep=",", row.names=FALSE, quote=FALSE)
```


### NEW RUN Mediators on HCC

```{r}
library(data.table)
f <- list.files(path="largedata/med_output", pattern="mediator_", full.names = T)
fmet <- list.files(path="largedata/med_output", pattern="mediator_X\\.", full.names = T)
fmic <- list.files(path="largedata/med_output", pattern="mediator_T[[0-9]", full.names = T)
id1 = which(f %in% fmet)
id2 = which(f %in% fmic)
f = f[-c(id1, id2)]

out <- data.frame()
for(i in 1:length(f)){
  tem <- read.csv(f[i], header=TRUE)
  tem$file <- f[i]
  out <- rbind(out, tem)
}

out$file  <- gsub(".*\\/", "", out$file)
out$method <- gsub("_.*", "", out$file)
out$tissue  <- gsub(".*_ts_|.csv", "", out$file)
out$trait  <- gsub(".*mediator_|_ts.*", "", out$file)
id = grep("P368", out$trait)
out = out [- id, ]

fwrite(out,"largedata/mediator_5100rows_agronomics.txt", sep="\t", row.names = F, quote = F)
```
### read in data metabolites on HCC

```{r}
library(data.table)
f <- list.files(path="largedata/med_output", pattern="mediator_X\\.", full.names = T)

out <- data.frame()
for(i in 1:length(f)){
  tem <- read.csv(f[i], header=TRUE)
  tem$file <- f[i]
  out <- rbind(out, tem)
}

out$file  <- gsub(".*\\/", "", out$file)
out$method <- gsub("_.*", "", out$file)
out$tissue  <- gsub(".*_ts_|.csv", "", out$file)
out$trait  <- gsub(".*mediator_|_ts.*", "", out$file)
fwrite(out,"largedata/mediator_13712rows_metabolites.txt", sep="\t", row.names = F, quote = F)
```


### Filter non-significant ones

```{r}
library(data.table)

ag_trait <- fread("data/geno_trait.txt")
m_trait <- fread("data/metabolite_rep1_genotype_namecorrect.txt")
trait_category <- data.frame(category = "other", trait = c(colnames(ag_trait)[-1], colnames(m_trait)[-1]))

trait_category$category[c(3,4,11:13,38)]= "Flowering Time"
trait_category$category[c(6,23)]= "Development"
trait_category$category[c(15,16,36)]= "Leaf Morphology"
trait_category$category[c(32,33)]= "Male Inflorescence"
trait_category$category[c(1,7,9,8)]= "Female Inflorescence"
trait_category$category[c(2,10,35,37,39,40)]= "Seed Traits"
trait_category$category[41:106]= "Metabolite"


#m <- read.csv("cache/mediators_23296rows.csv")
m1 <- fread("largedata/mediator_5100rows_agronomics.txt")
m2 <- fread("largedata/mediator_13712rows_metabolites.txt")
m = rbind(m1, m2)

trait_t = table(m$trait)
trait_t = as.data.frame(trait_t)

trait_c_t = table(trait_category$category)
trait_c_t = as.data.frame(trait_c_t)

m = merge(m, trait_category, by = "trait")

lin = c("Zm00001d030968", "Zm00001d049305", "Zm00001d039589", "Zm00001d038644", "Zm00001d036801", "Zm00001d034964", "Zm00001d020157", "Zm00001d013863", "Zm00001d042319", "Zm00001d033680", "Zm00001d045481", "Zm00001d006131", "Zm00001d006198", "Zm00001d032894", "Zm00001d017193", "Zm00001d011748", "Zm00001d010625", "Zm00001d006918", "Zm00001d034313", "Zm00001d049543", "Zm00001d017241", "Zm00001d006212", "Zm00001d021291", "Zm00001d045944", "Zm00001d034038", "Zm00001d033799", "Zm00001d028905", "Zm00001d013402" )
V3_V4 <- fread("./largedata/V3_V4.csv", header = FALSE , data.table=FALSE)
V3_V4 <- V3_V4[,c(1,2)]
Zea_mays.AGPv4_anno <- fread("./largedata/Zea_mays.AGPv4_anno.txt", header = TRUE, data.table = FALSE)
Zea_mays.AGPv4_anno <- data.frame(chr = Zea_mays.AGPv4_anno$chromosome_name, gene4 = Zea_mays.AGPv4_anno$ensembl_gene_id_v4, start4 = Zea_mays.AGPv4_anno$start_position, end4 = Zea_mays.AGPv4_anno$end_position, name = Zea_mays.AGPv4_anno$description)
Zea_mays.AGPv34_anno <- merge(V3_V4, Zea_mays.AGPv4_anno, by.x = "V2", by.y = "gene4")
Zea_mays.AGPv34_anno_s = subset(Zea_mays.AGPv34_anno, V2 %in% lin)
V3_V4_s = subset(V3_V4, V2 %in% lin)

which(flowering$id %in% V3_V4_s$V1)


oldm = fread("data/input/mediators_23296rows_by_cat.csv")
oldflowering = subset(oldm, category == "Flowering Time")
which(oldflowering$id %in% V3_V4_s$V1)

oldmads56 = subset(oldm, id == "GRMZM2G171650")

length(which(m$id %in% oldm$id))
#write.table(m, "data/input/mediators_23296rows_by_cat.csv", sep=",", row.names=FALSE, quote=FALSE)

write.table(m, "largedata/mediators_18812rows_by_cat.csv", sep=",", row.names=FALSE, quote=FALSE)

idx = which(m$id == "GRMZM2G171650")
mads56 = m[idx,c(1,2,6,9,10,11)]
mads56fixed = subset(mads56, method == "res.fixed")
mads56fixeds = mads56fixed[,c(6,1,5,3)]
library(dplyr)
mads56fixeds = arrange(mads56fixeds, category, trait, tissue, padj)
mads56fixeds_sig = subset(mads56fixeds, padj <= 0.05)


ms = m[,c(1,2,6,9,10,11)]
mfixed = subset(ms, method == "res.fixed")
mfixed = mfixed[,c(6,1,5,2,3)]
mfixed = mfixed[,c(4,1,2,3,5)]
mfixed = arrange(mfixed, id, category, trait, tissue, padj)

gene_freq = as.data.frame(table(mfixed$id))
gene_freq = arrange(gene_freq, desc(Freq))
gene_freq_g2 = subset(gene_freq, Freq >= 2)
idx = which(mfixed$id %in% gene_freq_g2$Var1)
mfixed_g2 = mfixed[idx, ]
write.table(mfixed_g2, "data/supplementary/mfixed_g2.csv", sep=",", row.names=FALSE, quote=FALSE)


mfixed_sig = subset(mfixed, padj <= 0.05)
siggene_freq = as.data.frame(table(mfixed_sig$id))
siggene_freq = arrange(siggene_freq, desc(Freq))


trait_medi_num = as.data.frame(table(mfixed$trait))
trait_medi_num = arrange(trait_medi_num, desc(Freq))

trait_sig_medi_num = as.data.frame(table(mfixed_sig$trait))
trait_sig_medi_num = arrange(trait_sig_medi_num, desc(Freq))

cat_trait_medinum = merge(trait_category, trait_medi_num, by.x = "trait", by.y = "Var1")
cat_trait_medinum = cat_trait_medinum[, c(2,1,3)]
cat_trait_medinum = arrange(cat_trait_medinum, category)
colnames(cat_trait_medinum)[3] = "num_mediators"
write.table(cat_trait_medinum, "data/supplementary/cat_trait_medinum.csv", sep=",", row.names=FALSE, quote=FALSE)

cat_trait_sig_medinum = merge(trait_category, trait_sig_medi_num, by.x = "trait", by.y = "Var1")
cat_trait_sig_medinum = cat_trait_sig_medinum[, c(2,1,3)]
cat_trait_sig_medinum = arrange(cat_trait_sig_medinum, category)
colnames(cat_trait_sig_medinum)[3] = "num_mediators"
write.table(cat_trait_sig_medinum, "data/supplementary/cat_trait_sig_medinum.csv", sep=",", row.names=FALSE, quote=FALSE)



for (i in 1:nrow(trait_c_t)) {
  idx = which(m$category == trait_c_t$Var1[i])
  fd = paste0("largedata/go/", trait_c_t$Var1[i], ".txt")
  write(unique(m$id[idx]), file = fd, sep = "\n")
}

Development = fread("largedata/go/Development_1.txt")

m5 <- subset(m, padj < 0.05)

for (i in 1:nrow(trait_c_t)) {
  idx = which(m5$category == trait_c_t$Var1[i])
  fd = paste0("largedata/go/", trait_c_t$Var1[i], "_p5.txt")
  write(unique(m5$id[idx]), file = fd, sep = "\n")
}

############READ GO RESULT METHOD 2 ############
f <- list.files(path="largedata/go/", pattern="_2", full.names = T)

out <- data.frame()
for(i in 1:length(f)){
  tem <- fread(f[i], header=TRUE)
  tem$file <- f[i]
  print(i)
  out <- rbind(out, tem)
}

out$file  <- gsub(".*\\/", "", out$file)
out$category  <- gsub("_2.*", "", out$file)

idx = grep("_p5", out$category)
go2 = out[-idx,]
gop52 = out[idx,]

gop52$category = gsub("_p5", "", gop52$category)

go2sig = subset(go2, FDR <0.05)
gop52sig = subset(gop52, FDR < 0.05)

############READ GO RESULT METHOD 2 ############

f <- list.files(path="largedata/go/", pattern="_1", full.names = T)

out <- data.frame()
for(i in 1:length(f)){
  tem <- fread(f[i], header=TRUE)
  tem$file <- f[i]
  print(i)
  out <- rbind(out, tem)
}

out$file  <- gsub(".*\\/", "", out$file)
out$category  <- gsub("_1.*", "", out$file)

idx = grep("_p5", out$category)
go1 = out[-idx,]
gop51 = out[idx,]

gop51$category = gsub("_p5", "", gop51$category)

go1sig = subset(go1, FDR <0.05)
gop51sig = subset(gop51, FDR < 0.05)

########################

library(ggplot2)


Development = subset(go1sig, category == "Development")

fsize =18
p1 = ggplot(Development, aes(x = Term, -log(pvalue))) +
    geom_bar(stat="identity", fill="#f68060", alpha=.6, width=.4) +
    coord_flip() +
    ggtitle("Development")+
    xlab("") +
    theme(plot.title = element_text(size=20, face = "bold"), 
          axis.text=element_text(size=16, face="bold"),
          strip.text.x = element_text(size = 16, face = "bold"),
          axis.title=element_text(size=fsize, face="bold"),
          legend.position = "right", 
          legend.title = element_text(size=fsize, face="bold"),
          legend.text = element_text(size=fsize))



Female_Inflorescence = subset(go1sig, category == "Female Inflorescence")


p2 = ggplot(Female_Inflorescence, aes(x = Term, y = -log(pvalue))) +
    geom_bar(stat="identity", fill="#f68060", alpha=.6, width=.4) +
    coord_flip() +
    ggtitle("Female Inflorescence")+
    xlab("") +
    theme(plot.title = element_text(size=20, face = "bold"), 
          axis.text=element_text(size=16, face="bold"),
          strip.text.x = element_text(size = 16, face = "bold"),
          axis.title=element_text(size=fsize, face="bold"),
          legend.position = "right", 
          legend.title = element_text(size=fsize, face="bold"),
          legend.text = element_text(size=fsize))


Flowering_Time = subset(go1sig, category == "Flowering Time")


p3 = ggplot(Flowering_Time, aes(x = Term, y=-log(pvalue))) +
    geom_bar(stat="identity", fill="#f68060", alpha=.6, width=.4) +
    coord_flip() +
    ggtitle("Flowering Time")+
    xlab("") +
    
    theme(plot.title = element_text(size=20, face = "bold"), 
          axis.text=element_text(size=14, face="bold"),
          strip.text.x = element_text(size = 14, face = "bold"),
          axis.title=element_text(size=fsize, face="bold"),
          legend.position = "right", 
          legend.title = element_text(size=fsize, face="bold"),
          legend.text = element_text(size=fsize))


Metabolite = subset(go1sig, category == "Metabolite")

p4 = ggplot(Metabolite, aes(x = Term, -log(pvalue))) +
    geom_bar(stat="identity", fill="#f68060", alpha=.6, width=.4) +
    coord_flip() +
    ggtitle("Metabolite")+
    xlab("") +
    theme(plot.title = element_text(size=20, face = "bold"), 
          axis.text=element_text(size=10, face="bold"),
          strip.text.x = element_text(size = 10, face = "bold"),
          axis.title=element_text(size=fsize, face="bold"),
          legend.position = "right", 
          legend.title = element_text(size=fsize, face="bold"),
          legend.text = element_text(size=fsize))


other = subset(go1sig, category == "other")


p5 = ggplot(other, aes(x = Term, y=-log(pvalue))) +
    geom_bar(stat="identity", fill="#f68060", alpha=.6, width=.4) +
    coord_flip() +
    ggtitle("other")+
    xlab("") +
    
    theme(plot.title = element_text(size=20, face = "bold"), 
          axis.text=element_text(size=14, face="bold"),
          strip.text.x = element_text(size = 14, face = "bold"),
          axis.title=element_text(size=fsize, face="bold"),
          legend.position = "right", 
          legend.title = element_text(size=fsize, face="bold"),
          legend.text = element_text(size=fsize))

Seed_Traits = subset(go1sig, category == "Seed Traits")


p6 = ggplot(Seed_Traits, aes(x = Term, y=-log(pvalue))) +
    geom_bar(stat="identity", fill="#f68060", alpha=.6, width=.4) +
    coord_flip() +
    ggtitle("Seed Traits")+
    xlab("") +
    
    theme(plot.title = element_text(size=20, face = "bold"), 
          axis.text=element_text(size=14, face="bold"),
          strip.text.x = element_text(size = 14, face = "bold"),
          axis.title=element_text(size=fsize, face="bold"),
          legend.position = "right", 
          legend.title = element_text(size=fsize, face="bold"),
          legend.text = element_text(size=fsize))

Male_Inflorescence = subset(go1sig, category == "Male Inflorescence")


p7 = ggplot(Male_Inflorescence, aes(x = Term, y = -log(pvalue))) +
    geom_bar(stat="identity", fill="#f68060", alpha=.6, width=.4) +
    coord_flip() +
    ggtitle("Male Inflorescence")+
    xlab("") +
    theme(plot.title = element_text(size=20, face = "bold"), 
          axis.text=element_text(size=12, face="bold"),
          strip.text.x = element_text(size = 12, face = "bold"),
          axis.title=element_text(size=fsize, face="bold"),
          legend.position = "right", 
          legend.title = element_text(size=fsize, face="bold"),
          legend.text = element_text(size=fsize))

library("ggpubr")
library(gridExtra)
ggarrange(p2,p7, nrow =1, ncol = 2)

#####0.05 significant visual####
Flowering_Time = subset(gop51sig, category == "Flowering Time")


p8 = ggplot(Flowering_Time, aes(x = Term, y=-log(pvalue))) +
    geom_bar(stat="identity", fill="#f68060", alpha=.6, width=.4) +
    coord_flip() +
    ggtitle("Flowering Time")+
    xlab("") +
    
    theme(plot.title = element_text(size=20, face = "bold"), 
          axis.text=element_text(size=14, face="bold"),
          strip.text.x = element_text(size = 14, face = "bold"),
          axis.title=element_text(size=fsize, face="bold"),
          legend.position = "right", 
          legend.title = element_text(size=fsize, face="bold"),
          legend.text = element_text(size=fsize))


Metabolite = subset(gop51sig, category == "Metabolite")

p9 = ggplot(Metabolite, aes(x = Term, -log(pvalue))) +
    geom_bar(stat="identity", fill="#f68060", alpha=.6, width=.4) +
    coord_flip() +
    ggtitle("Metabolite")+
    xlab("") +
    theme(plot.title = element_text(size=20, face = "bold"), 
          axis.text=element_text(size=16, face="bold"),
          strip.text.x = element_text(size = 16, face = "bold"),
          axis.title=element_text(size=fsize, face="bold"),
          legend.position = "right", 
          legend.title = element_text(size=fsize, face="bold"),
          legend.text = element_text(size=fsize))


Leaf_Morphology = subset(gop51sig, category == "Leaf Morphology")

p10 = ggplot(Leaf_Morphology, aes(x = Term, -log(pvalue))) +
    geom_bar(stat="identity", fill="#f68060", alpha=.6, width=.4) +
    coord_flip() +
    ggtitle("Leaf Morphology")+
    xlab("") +
    theme(plot.title = element_text(size=20, face = "bold"), 
          axis.text=element_text(size=16, face="bold"),
          strip.text.x = element_text(size = 16, face = "bold"),
          axis.title=element_text(size=fsize, face="bold"),
          legend.position = "right", 
          legend.title = element_text(size=fsize, face="bold"),
          legend.text = element_text(size=fsize))





# nrow(m5) 2751
m1 <- subset(m, padj < 0.01)
# 1567

t5 <- as.data.frame(table(m5$id))
t5 <- t5[order(t5$Freq, decreasing=T),]
```

focus on only the flowering time traits

```{r}
mflowering <- subset(m5, trait %in% c("DaystoSilk",	"DaysToTassel", "GDDDaystoSilk",	"GDDDaystoTassel")) 

mflowering <- subset(mflowering, method == "res.fixed") 

male_mflowering <- subset(mflowering, trait %in% c(	"DaysToTassel",	"GDDDaystoTassel")) 
female_mflowering <- subset(mflowering, trait %in% c(	"DaystoSilk",	"GDDDaystoSilk")) 

tp <- as.data.frame(table(mt$id))
tp <- tp[order(tp$Freq, decreasing = T),]

library("readxl")
maleg<- read_excel("largedata/top_genes_associated_with_male_flowering.xlsx")
femaleg<- read_excel("largedata/top_genes_associated_with_female_flowering.xlsx")

which(male_mflowering$id %in% maleg$ID)
which(female_mflowering$id %in% femaleg$ID)

length(which(male_mflowering$id %in% maleg$ID))



Nf <- nrow(femaleg)
N <- 37127

noverlap <-length(which(female_mflowering$id %in% femaleg$ID))
n <- nrow(female_mflowering)

p <- Nf/N
obsfreq <- c(noverlap, n - noverlap)
nullprobs <- c(p, 1-p)
chisq.test(obsfreq,p=nullprobs)

```
```{r}
flowering = subset(m, category == "Flowering Time")
mflowering <- subset(m, trait %in% c("DaystoSilk",	"DaysToTassel", "GDDDaystoSilk",	"GDDDaystoTassel")) 

mflowering <- subset(mflowering, method == "res.fixed") 

male_mflowering <- subset(mflowering, trait %in% c(	"DaysToTassel",	"GDDDaystoTassel")) 
female_mflowering <- subset(mflowering, trait %in% c(	"DaystoSilk",	"GDDDaystoSilk")) 

tp <- as.data.frame(table(mt$id))
tp <- tp[order(tp$Freq, decreasing = T),]

library("readxl")
maleg<- read_excel("largedata/top_genes_associated_with_male_flowering.xlsx")
femaleg<- read_excel("largedata/top_genes_associated_with_female_flowering.xlsx")

which(male_mflowering$id %in% maleg$ID)
which(female_mflowering$id %in% femaleg$ID)

length(which(male_mflowering$id %in% maleg$ID))



Nf <- nrow(femaleg)
N <- 37127

noverlap <-length(which(female_mflowering$id %in% femaleg$ID))
n <- nrow(female_mflowering)

p <- Nf/N
obsfreq <- c(noverlap, n - noverlap)
nullprobs <- c(p, 1-p)
chisq.test(obsfreq,p=nullprobs)


```


### Mediators

```{r}

f368 <- list.files(path="/common/jyanglab/zhikaiyang/projects/mediation/largedata/med_output/", pattern="mediator_P368", full.names = T)

out <- data.frame()
for(i in 1:length(f368)){
  tem <- read.csv(f368[i], header=TRUE)
  tem$file <- f368[i]
  out <- rbind(out, tem)
}

out$file  <- gsub(".*\\/", "", out$file)
# method
out$method <- gsub("_.*", "", out$file)
# tissue type
out$tissue <- gsub(".*_ts_P368\\.|.csv", "", out$file)
table(out$tissue)
# trait
out$trait <- gsub(".*mediator_P368\\.|_ts.*", "", out$file)
table(out$trait)

write.table(out, "cache/mediators_P368.csv", sep=",", row.names=FALSE, quote=FALSE)
```


### Filter non-significant ones

```{r}
m_kern <- subset(m, tissue == "Kern")

m_368 <- read.csv("cache/mediators_P368.csv")

intersect(m_kern$id[which(m_kern$method == "res.eq")], m_368$id[which(m_368$method == "res.eq")])
#[1] "GRMZM2G305834" "GRMZM2G179679"
intersect(m_kern$id[which(m_kern$method == "res.fixed")], m_368$id[which(m_368$method == "res.fixed")])
#character(0)
GRMZM2G305834_kern <- subset(m_kern[which(m_kern$method == "res.eq"),], id == "GRMZM2G305834")
GRMZM2G305834_368 <- subset(m_368[which(m_368$method == "res.eq"),], id == "GRMZM2G305834")


m5_368 <- subset(m_368, padj < 0.05)
# nrow(m5) 2751
m1_368 <- subset(m_368, padj < 0.01)
# 1567

t5_368 <- as.data.frame(table(m5_368$id))
t5_368 <- t5_368[order(t5_368$Freq, decreasing=T),]
```


