---
title: "Collect mediation analysis results"
author: "Zhikai Yang "
date: "11-2-2020"
output: NULL
---


## Normalize the path:

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE) 
knitr::opts_knit$set(root.dir=normalizePath('../../')) 
```

### Results generated by Zhikai

/common/jyanglab/zhikaiyang/projects/mediation/largedata/med_output

fixed_dsnps : direct snps
fixed_indirect : indirect snps
fixed_mediator_trait : mediator for agronomic traits
fixed_mediator_X...: mediator for metabolic traits
fixe.pcut_sum : pvm


### on HCC

```{r}

f <- list.files(path="/common/jyanglab/zhikaiyang/projects/mediation/largedata/med_output/", pattern="_dsnps_", full.names = T)

out <- data.frame()
for(i in 1:length(f)){
  tem <- read.csv(f[i], header=TRUE)
  tem$file <- f[i]
  out <- rbind(out, tem)
}

out$file  <- gsub(".*\\/", "", out$file)

library(data.table)

snps_p7msnps <- fread("/common/jyanglab/zhikaiyang/projects/mediation-analysis/largedata/code2020/gen/allchr_bisnp_n282_snpid_maf01_geno2_pruned.bim", header = FALSE , data.table=FALSE)
out <- merge(out, snps_p7msnps[,c(1,2,4)], by.x = "snp", by.y = "V2", sort = F, all.x = T)
colnames(out)[c(5,6)] = c("chr", "pos")
# method
out$method <- gsub("_.*", "", out$file)
# tissue type
out$tissue <- gsub(".*_ts_|.csv", "", out$file)
table(out$tissue)
# trait
out$trait <- gsub(".*dsnps_|_ts.*", "", out$file)
table(out$trait)

write.table(out, "largedata/dsnps_vs_gwas/dsnps_20149rows.csv", sep=",", row.names=FALSE, quote=FALSE)
```


### on local

```{r}
dsnps <- fread("largedata/dsnps_vs_gwas/dsnps_20149rows.csv")
all_gwas <- fread("largedata/dsnps_vs_gwas/all_traits.all_res.txt")
which(all_gwas$Chr == dsnps$chr[1] & all_gwas$start < dsnps$pos[1] & all_gwas$end > dsnps$pos[1] & all_gwas$date == dsnps$trait[1])
idx <- grep("X\\.", all_gwas$date)
all_gwas$date[idx] <- gsub("\\.$", "", all_gwas$date[idx])
dsnps$gwas_overlap <- 0

idx <- grep("P368\\.", dsnps$trait)
dsnps = dsnps[- idx,]

for (i in 1:nrow(dsnps)) {
  if (length(which(all_gwas$Chr == dsnps$chr[i] & all_gwas$start < dsnps$pos[i] & all_gwas$end > dsnps$pos[i] & all_gwas$date == dsnps$trait[i])) != 0) {
    dsnps$gwas_overlap[i] = which(all_gwas$Chr == dsnps$chr[i] & all_gwas$start < dsnps$pos[i] & all_gwas$end > dsnps$pos[i] & all_gwas$date == dsnps$trait[i])
  }
  
}

write.table(dsnps, "largedata/dsnps_vs_gwas/dsnps_100kb_comparsion.csv", sep=",", row.names=FALSE, quote=FALSE)

dsnps <- fread("largedata/dsnps_vs_gwas/dsnps_100kb_comparsion.csv")

dsnps_eq <- subset(dsnps, method == "res.eq")
dsnps_fixed <- subset(dsnps, method == "res.fixed")
i_trait_eq <- unique(dsnps_eq$trait)
i_tissue_eq <- unique(dsnps_eq$tissue)
gwas_overlap_eq = NULL
for (trait in i_trait_eq) {
  
  for (tissue in i_tissue_eq) {
    t = intersect(which(dsnps_eq$trait == trait ), which( dsnps_eq$tissue == tissue))
    if (length(t) > 0) {
      tem = dsnps_eq[t, ]
      o_ratio = length(which(tem$gwas_overlap != 0))/nrow(tem)
      gwas_overlap_eq = rbind(gwas_overlap_eq, c(trait, tissue, o_ratio, length(which(tem$gwas_overlap != 0)), nrow(tem) ))
    }
    
  }
  
}


i_trait_fixed <- unique(dsnps_fixed$trait)
i_tissue_fixed <- unique(dsnps_fixed$tissue)
gwas_overlap_fixed = NULL
for (trait in i_trait_fixed) {
  
  for (tissue in i_tissue_fixed) {
    t = intersect(which(dsnps_fixed$trait == trait ), which( dsnps_fixed$tissue == tissue))
    if (length(t) > 0) {
      tem = dsnps_fixed[t, ]
      o_ratio = length(which(tem$gwas_overlap != 0))/nrow(tem)
      gwas_overlap_fixed = rbind(gwas_overlap_fixed, c(trait, tissue, o_ratio, length(which(tem$gwas_overlap != 0)), nrow(tem) ))
    }
    
  }
  
}
```

```{r}
ag_trait <- fread("data/geno_trait.txt")
m_trait <- fread("data/metabolite_rep1_genotype_namecorrect.txt")
trait_category <- data.frame(category = "other", trait = c(colnames(ag_trait)[-1], colnames(m_trait)[-1]))

trait_category$category[c(3,4,11:13,38)]= "Flowering Time"
trait_category$category[c(6,23)]= "Development"
trait_category$category[c(15,16,36)]= "Leaf Morphology"
trait_category$category[c(32,33)]= "Male Inflorescence"
trait_category$category[c(1,7,9,8)]= "Female Inflorescence"
trait_category$category[c(2,10,35,37,39,40)]= "Seed Traits"
trait_category$category[41:106]= "Metabolite"

dsnps = merge(dsnps, trait_category, by = "trait", all.x = T, sort = F)
dsnps_eq <- subset(dsnps, method == "res.eq")
dsnps_fixed <- subset(dsnps, method == "res.fixed")

gwas_overlap_eq = as.data.frame(gwas_overlap_eq)
colnames(gwas_overlap_eq) = c("trait", "tissue", "o_ratio", "num_o", "N_dsnps")
gwas_overlap_fixed = as.data.frame(gwas_overlap_fixed)
colnames(gwas_overlap_fixed) = c("trait", "tissue", "o_ratio", "num_o", "N_dsnps")

gwas_overlap_eq = merge(gwas_overlap_eq, trait_category, by = "trait", all.x = T, sort = F)
gwas_overlap_fixed = merge(gwas_overlap_fixed, trait_category, by = "trait", all.x = T, sort = F)

```

### all plot
```{r}
fsize=18
gwas_overlap_eq$tissue <- factor(gwas_overlap_eq$tissue, levels = c( "GRoot", "GShoot", "L3Base", "L3Tip", "LMAD", "LMAN","Kern"),
                    labels =c( "Germinating Root", "Germinating Shoot", "Leaf 3 base", "Leaf 3 tip", "Adult leaf during the day", "Adult leaf at night","Kernels after pollination"))
gwas_overlap_eq$o_ratio = as.numeric(gwas_overlap_eq$o_ratio)

library(ggplot2)

#df0 <- subset(df, cat != "1")
pall <- ggplot(gwas_overlap_eq, aes(x=tissue, y=o_ratio, fill=category)) +
    #geom_bar(stat="identity", position=position_dodge()) +
    geom_boxplot(outlier.colour="black", outlier.shape=16, 
             outlier.size=1, notch=FALSE) +
    # scale_fill_manual(values=c("#69b3a2", "#cdc0b0")) +
    # scale_fill_viridis(discrete = TRUE, alpha=0.6) +
    # geom_jitter(color="#cc4141", size=2, alpha=0.9) +
    xlab("") +
    ylab("Overlap ratio with GWAS signals") +
    ggtitle("") +
    theme_classic() +
    labs(fill = "") +
    scale_y_continuous(limits = c(-0.01, 1)) +
    theme(plot.title = element_text(size=20, face = "bold"), 
          axis.text=element_text(size=16, face="bold"),
          strip.text.x = element_text(size = 16, face = "bold"),
          axis.title=element_text(size=fsize, face="bold"),
          axis.text.x = element_text(angle = 15, hjust=0.8),
          legend.position = "right", 
          legend.title = element_text(size=fsize, face="bold"),
          legend.text = element_text(size=fsize))
pall

pdf("graphs/overlap_ratio_gwas_100kb_all_traits_eq.pdf", width = 9, height= 6)
pall
dev.off()

```

### all plot
```{r}
fsize=18
gwas_overlap_fixed$tissue <- factor(gwas_overlap_fixed$tissue, levels = c( "GRoot", "GShoot", "L3Base", "L3Tip", "LMAD", "LMAN","Kern"),
                    labels =c( "Germinating Root", "Germinating Shoot", "Leaf 3 base", "Leaf 3 tip", "Adult leaf during the day", "Adult leaf at night","Kernels after pollination"))
gwas_overlap_fixed$o_ratio = as.numeric(gwas_overlap_fixed$o_ratio)

library(ggplot2)

#df0 <- subset(df, cat != "1")
pall <- ggplot(gwas_overlap_fixed, aes(x=tissue, y=o_ratio, fill=category)) +
    #geom_bar(stat="identity", position=position_dodge()) +
    geom_boxplot(outlier.colour="black", outlier.shape=16, 
             outlier.size=1, notch=FALSE) +
    # scale_fill_manual(values=c("#69b3a2", "#cdc0b0")) +
    # scale_fill_viridis(discrete = TRUE, alpha=0.6) +
    # geom_jitter(color="#cc4141", size=2, alpha=0.9) +
    xlab("") +
    ylab("Overlap ratio with GWAS signals") +
    ggtitle("") +
    theme_classic() +
    labs(fill = "") +
    scale_y_continuous(limits = c(-0.01, 1)) +
    theme(plot.title = element_text(size=20, face = "bold"), 
          axis.text=element_text(size=16, face="bold"),
          strip.text.x = element_text(size = 16, face = "bold"),
          axis.title=element_text(size=fsize, face="bold"),
          axis.text.x = element_text(angle = 15, hjust=0.8),
          legend.position = "right", 
          legend.title = element_text(size=fsize, face="bold"),
          legend.text = element_text(size=fsize))
pall

pdf("graphs/overlap_ratio_gwas_100kb_all_traits_fixed.pdf", width = 9, height= 6)
pall
dev.off()

```


### enrichment test

```{r}
index = unique(all_gwas$date)

gwas_nsig = NULL
for (trait in index) {
  idx = which(all_gwas$date == trait)
  tem = all_gwas[idx,c(1,7)]
  gwas_nsig = rbind(gwas_nsig, c(trait, sum(tem$N), 20724054))
}

gwas_nsig = as.data.frame(gwas_nsig)
colnames(gwas_nsig) = c("trait", "N_sig", "N_total")
idx = grep("_368",  gwas_nsig$trait)
gwas_nsig = gwas_nsig[-idx,]
gwas_nsig$N_sig = as.numeric(gwas_nsig$N_sig)
gwas_nsig$N_total = as.numeric(gwas_nsig$N_total)
gwas_nsig$p_sig = gwas_nsig$N_sig/gwas_nsig$N_total

gwas_overlap_eq = merge(gwas_overlap_eq, gwas_nsig, by = "trait", all.x = T, sort = F)
gwas_overlap_fixed = merge(gwas_overlap_fixed, gwas_nsig, by = "trait", all.x = T, sort = F)

gwas_overlap_eq$num_o = as.numeric(gwas_overlap_eq$num_o)
gwas_overlap_fixed$num_o = as.numeric(gwas_overlap_fixed$num_o)
gwas_overlap_eq$N_dsnps = as.numeric(gwas_overlap_eq$N_dsnps)
gwas_overlap_fixed$N_dsnps = as.numeric(gwas_overlap_fixed$N_dsnps)

gwas_overlap_eq$p_val_chi = 0.0
gwas_overlap_eq$p_val_chi_adj = 0.0
gwas_overlap_eq$statistic_chi = 0.0
for (i in 1:nrow(gwas_overlap_eq)) {
  t = chisq.test(c(gwas_overlap_eq$num_o[i], gwas_overlap_eq$N_dsnps[i]-gwas_overlap_eq$num_o[i]),p=c(gwas_overlap_eq$p_sig[i], 1-gwas_overlap_eq$p_sig[i]))
  gwas_overlap_eq$statistic_chi[i] = t$statistic
  gwas_overlap_eq$p_val_chi[i] = t$p.value
  gwas_overlap_eq$p_val_chi_adj[i] = p.adjust(gwas_overlap_eq$p_val_chi[i], method = "fdr", n= nrow(gwas_overlap_eq))
}

gwas_overlap_fixed$p_val_chi = 0.0
gwas_overlap_fixed$p_val_chi_adj = 0.0
gwas_overlap_fixed$statistic_chi = 0.0
for (i in 1:nrow(gwas_overlap_fixed)) {
  t = chisq.test(c(gwas_overlap_fixed$num_o[i], gwas_overlap_fixed$N_dsnps[i]-gwas_overlap_fixed$num_o[i]),p=c(gwas_overlap_fixed$p_sig[i], 1-gwas_overlap_fixed$p_sig[i]))
  gwas_overlap_fixed$statistic_chi[i] = t$statistic
  gwas_overlap_fixed$p_val_chi[i] = t$p.value
  gwas_overlap_fixed$p_val_chi_adj[i] = p.adjust(gwas_overlap_fixed$p_val_chi[i], method = "fdr", n= nrow(gwas_overlap_fixed))
}

length(which(gwas_overlap_eq$p_val_chi_adj > 0.05))
length(which(gwas_overlap_fixed$p_val_chi_adj > 0.05))



hist(gwas_overlap_eq$p_val_chi_adj[- which(gwas_overlap_eq$p_val_chi_adj > 0.05)], main = "Distribution of adjusted p-value by FDR (Medfix.0.5)", xlab = "adjusted p-value")

hist(gwas_overlap_fixed$p_val_chi_adj[- which(gwas_overlap_fixed$p_val_chi_adj > 0.05)], main = "Distribution of adjusted p-value by FDR (Medfix.bic)", xlab = "adjusted p-value")

write.table(gwas_overlap_eq, "largedata/dsnps_vs_gwas/gwas_100kb_overlap_eq.csv", sep=",", row.names=FALSE, quote=FALSE)
write.table(gwas_overlap_fixed, "largedata/dsnps_vs_gwas/gwas_100kb_overlap_fixed.csv", sep=",", row.names=FALSE, quote=FALSE)


```

